repose-aggregator/src/docs/asciidoc/services/datastores.adoc

```diff
@@ -3,4 +3,293 @@
+
+`GET /powerapi/dist-datastore/objects/object-key`
+
+becomes
+
+`GET /powerapi/dist-datastore/objects/cecda330-5a61-26cd-1a71-d5fe34a8e302`
+
+Obscuring key-space is not a function of the distributed datastore service.
+This functionality is only present in the internally consumed java cache client.
+If an external application puts an object into the distributed datastore, the object will be stored under the value of the key given.
+
+=== Remote Management
+The repose distributed datastore component is a filter component that hosts a simple RESTful API that can be contacted to perform remote object store operations.
+These operations are defined below.
+
+`GET /powerapi/dist-datastore/objects/<object-key> HTTP/1.1`
+
+Gets a stored object from the datastore by its key.
+
+`PUT /powerapi/dist-datastore/objects/<object-key> HTTP/1.1`
+
+Puts an object into the datastore by its key.
+
+`DELETE /powerapi/dist-datastore/objects/<object-key> HTTP/1.1`
+
+Deletes an object from the datastore by its key.
+
+`PATCH /powerapi/dist-datastore/objects/<object-key> HTTP/1.1`
+
+Patches an object in the datastore by its key.
+If the object does not exist, then a new one created.
+Return the modified/new object.
+The object must be Patchable.
+
+=== Remote Fail-Over
+In the event that a node with in a datastore cluster falls off line or is unable to respond to requests, it is removed from the node's cluster membership for a period of time.
+During this time, the online node will then re-address its key-space in order to continue operation.
+After certain periods of rest, the node may attempt to introduce the damaged cluster member into its cluster membership.
+A damaged cluster member must go through several validation passes where the member is introduced back into the addressing algorithm before it can be considered online.
+In order to keep healthy nodes from attempting to route requests to the damaged node, a participating node may tell it's destination that the destination may not route the request and must store, delete or get the value locally.
+
+=== Performance
+The Repose node will open sockets each time it has to communicate with other Repose nodes to share information.
+During times of load this can affect performance and data integrity as when one node cannot communicate with another it will mark that node damaged and store/create information locally.
+One way this can happen is if the user running repose hits their open file limit.
+Luckily this can be mitigated by increasing the open file limit for the user running Repose.
+
+=== JMX Reporting
+Currently Repose instances do not report Distributed Datastore information to JMX.
+This is something that has been done in the past, but an upgrade to the metrics library used has made this capability incompatible with the current codebase.
+
+=== Authentication
+The authentication components will use the local datastore to cache token, role, and group information.
```

Comment:
This seems like it doesn't really belong here. Sure, authentication components consume the datastore service, but increasingly so do other components. This has nothing to do with the datastore service, which is data-agnostic, and everything to do with the client component. Therefore, I feel like this should be called out on those other components' respective pages. Same with the rate limiting bit below.